{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Gene Expression RNA-Seq Classification \n",
    " \n",
    "#### Classification Models\n",
    "I will be working with the subsets of my original normalized dataframe that only contains genes with a certain correlation.\n",
    "\n",
    "The three correlation threshold I will work with are: \n",
    "- 0.262105\n",
    "- 0.50\n",
    "- 0.75\n",
    "\n",
    "These will be the datasets I will work with first to explore different models and then I will re-adjust my datasets if needed to terst out different models. "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "There are different types of classification models:\n",
    "- Binary\n",
    "- Multi-Class\n",
    "- Multi-Label\n",
    "- Imbalanced Classifications\n",
    "\n",
    "I am currently exploring various classification models for my dataset, which primarily aligns with Multi-Class Classification. I aim to determine the most suitable algorithm among the available Multi-Class options.\n",
    "\n",
    "In addition to Multi-Class Classification, I am intrigued by the idea of testing the dataset with a Multi-Label algorithm. Given that my dataset involves gene expression levels and tumor classification, I am particularly interested in understanding how the model assigns percentages to specific tumor types compared to others. This approach will provide insights into the nuanced relationships between gene expressions and different tumor classifications."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Import packages \n",
    "import pandas as pd"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### 1. Random Forest Algorithm"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Unnamed: 0</th>\n",
       "      <th>Class</th>\n",
       "      <th>class_encoded</th>\n",
       "      <th>gene_17801</th>\n",
       "      <th>gene_89</th>\n",
       "      <th>gene_19339</th>\n",
       "      <th>gene_16372</th>\n",
       "      <th>gene_6611</th>\n",
       "      <th>gene_14092</th>\n",
       "      <th>gene_15906</th>\n",
       "      <th>gene_1750</th>\n",
       "      <th>gene_18745</th>\n",
       "      <th>gene_1784</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>sample_0</td>\n",
       "      <td>PRAD</td>\n",
       "      <td>4</td>\n",
       "      <td>0.739927</td>\n",
       "      <td>0.808553</td>\n",
       "      <td>0.283090</td>\n",
       "      <td>0.694543</td>\n",
       "      <td>0.735059</td>\n",
       "      <td>0.357113</td>\n",
       "      <td>0.724501</td>\n",
       "      <td>0.134339</td>\n",
       "      <td>0.741183</td>\n",
       "      <td>0.613634</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>sample_1</td>\n",
       "      <td>LUAD</td>\n",
       "      <td>3</td>\n",
       "      <td>0.451757</td>\n",
       "      <td>0.515478</td>\n",
       "      <td>0.343139</td>\n",
       "      <td>0.434428</td>\n",
       "      <td>0.523506</td>\n",
       "      <td>0.242179</td>\n",
       "      <td>0.306045</td>\n",
       "      <td>0.282810</td>\n",
       "      <td>0.119347</td>\n",
       "      <td>0.511011</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>sample_2</td>\n",
       "      <td>PRAD</td>\n",
       "      <td>4</td>\n",
       "      <td>0.090283</td>\n",
       "      <td>0.853200</td>\n",
       "      <td>0.054013</td>\n",
       "      <td>0.626688</td>\n",
       "      <td>0.244987</td>\n",
       "      <td>0.361074</td>\n",
       "      <td>0.145850</td>\n",
       "      <td>0.090265</td>\n",
       "      <td>0.589364</td>\n",
       "      <td>0.568915</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>sample_3</td>\n",
       "      <td>PRAD</td>\n",
       "      <td>4</td>\n",
       "      <td>0.486308</td>\n",
       "      <td>0.926782</td>\n",
       "      <td>0.386205</td>\n",
       "      <td>0.950615</td>\n",
       "      <td>0.501546</td>\n",
       "      <td>0.399962</td>\n",
       "      <td>0.449855</td>\n",
       "      <td>0.123032</td>\n",
       "      <td>0.930190</td>\n",
       "      <td>0.926884</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>sample_4</td>\n",
       "      <td>BRCA</td>\n",
       "      <td>0</td>\n",
       "      <td>0.120691</td>\n",
       "      <td>0.353085</td>\n",
       "      <td>0.826373</td>\n",
       "      <td>0.423637</td>\n",
       "      <td>0.269736</td>\n",
       "      <td>0.894639</td>\n",
       "      <td>0.200712</td>\n",
       "      <td>0.654923</td>\n",
       "      <td>0.195289</td>\n",
       "      <td>0.252237</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>796</th>\n",
       "      <td>sample_796</td>\n",
       "      <td>BRCA</td>\n",
       "      <td>0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.346446</td>\n",
       "      <td>0.862536</td>\n",
       "      <td>0.196768</td>\n",
       "      <td>0.115289</td>\n",
       "      <td>0.837581</td>\n",
       "      <td>0.043063</td>\n",
       "      <td>0.573846</td>\n",
       "      <td>0.092341</td>\n",
       "      <td>0.268784</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>797</th>\n",
       "      <td>sample_797</td>\n",
       "      <td>LUAD</td>\n",
       "      <td>3</td>\n",
       "      <td>0.428152</td>\n",
       "      <td>0.579990</td>\n",
       "      <td>0.134291</td>\n",
       "      <td>0.675421</td>\n",
       "      <td>0.536195</td>\n",
       "      <td>0.273634</td>\n",
       "      <td>0.119668</td>\n",
       "      <td>0.259258</td>\n",
       "      <td>0.123537</td>\n",
       "      <td>0.589349</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>798</th>\n",
       "      <td>sample_798</td>\n",
       "      <td>COAD</td>\n",
       "      <td>1</td>\n",
       "      <td>0.728309</td>\n",
       "      <td>0.546568</td>\n",
       "      <td>0.502283</td>\n",
       "      <td>0.576902</td>\n",
       "      <td>0.856168</td>\n",
       "      <td>0.489486</td>\n",
       "      <td>0.539400</td>\n",
       "      <td>0.564654</td>\n",
       "      <td>0.293789</td>\n",
       "      <td>0.698904</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>799</th>\n",
       "      <td>sample_799</td>\n",
       "      <td>PRAD</td>\n",
       "      <td>4</td>\n",
       "      <td>0.724511</td>\n",
       "      <td>0.844504</td>\n",
       "      <td>0.399249</td>\n",
       "      <td>0.879369</td>\n",
       "      <td>0.751820</td>\n",
       "      <td>0.434896</td>\n",
       "      <td>0.653873</td>\n",
       "      <td>0.289992</td>\n",
       "      <td>0.829432</td>\n",
       "      <td>0.774361</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>800</th>\n",
       "      <td>sample_800</td>\n",
       "      <td>PRAD</td>\n",
       "      <td>4</td>\n",
       "      <td>0.684042</td>\n",
       "      <td>0.723963</td>\n",
       "      <td>0.379439</td>\n",
       "      <td>0.637061</td>\n",
       "      <td>0.722141</td>\n",
       "      <td>0.586010</td>\n",
       "      <td>0.618162</td>\n",
       "      <td>0.104089</td>\n",
       "      <td>0.307389</td>\n",
       "      <td>0.749187</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>801 rows Ã— 13 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "     Unnamed: 0 Class  class_encoded  gene_17801   gene_89  gene_19339  \\\n",
       "0      sample_0  PRAD              4    0.739927  0.808553    0.283090   \n",
       "1      sample_1  LUAD              3    0.451757  0.515478    0.343139   \n",
       "2      sample_2  PRAD              4    0.090283  0.853200    0.054013   \n",
       "3      sample_3  PRAD              4    0.486308  0.926782    0.386205   \n",
       "4      sample_4  BRCA              0    0.120691  0.353085    0.826373   \n",
       "..          ...   ...            ...         ...       ...         ...   \n",
       "796  sample_796  BRCA              0    0.000000  0.346446    0.862536   \n",
       "797  sample_797  LUAD              3    0.428152  0.579990    0.134291   \n",
       "798  sample_798  COAD              1    0.728309  0.546568    0.502283   \n",
       "799  sample_799  PRAD              4    0.724511  0.844504    0.399249   \n",
       "800  sample_800  PRAD              4    0.684042  0.723963    0.379439   \n",
       "\n",
       "     gene_16372  gene_6611  gene_14092  gene_15906  gene_1750  gene_18745  \\\n",
       "0      0.694543   0.735059    0.357113    0.724501   0.134339    0.741183   \n",
       "1      0.434428   0.523506    0.242179    0.306045   0.282810    0.119347   \n",
       "2      0.626688   0.244987    0.361074    0.145850   0.090265    0.589364   \n",
       "3      0.950615   0.501546    0.399962    0.449855   0.123032    0.930190   \n",
       "4      0.423637   0.269736    0.894639    0.200712   0.654923    0.195289   \n",
       "..          ...        ...         ...         ...        ...         ...   \n",
       "796    0.196768   0.115289    0.837581    0.043063   0.573846    0.092341   \n",
       "797    0.675421   0.536195    0.273634    0.119668   0.259258    0.123537   \n",
       "798    0.576902   0.856168    0.489486    0.539400   0.564654    0.293789   \n",
       "799    0.879369   0.751820    0.434896    0.653873   0.289992    0.829432   \n",
       "800    0.637061   0.722141    0.586010    0.618162   0.104089    0.307389   \n",
       "\n",
       "     gene_1784  \n",
       "0     0.613634  \n",
       "1     0.511011  \n",
       "2     0.568915  \n",
       "3     0.926884  \n",
       "4     0.252237  \n",
       "..         ...  \n",
       "796   0.268784  \n",
       "797   0.589349  \n",
       "798   0.698904  \n",
       "799   0.774361  \n",
       "800   0.749187  \n",
       "\n",
       "[801 rows x 13 columns]"
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Get our training datasets \n",
    "subset_data_1 = pd.read_csv(\"/Users/kim/Desktop/repos/RNA-Seq_GeneExpression_Model/subset_min-max_threshold_df_0.26.csv\")\n",
    "subset_data_2 = pd.read_csv(\"/Users/kim/Desktop/repos/RNA-Seq_GeneExpression_Model/subset_min-max_threshold_df_0.50.csv\")\n",
    "subset_data_3 = pd.read_csv(\"/Users/kim/Desktop/repos/RNA-Seq_GeneExpression_Model/subset_min-max_threshold_df_0.75.csv\")"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "base",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.13"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
